services/src/main/java/io/grpc/services/ChannelzService.java

```diff
@@ -0,0 +1,167 @@
+package io.grpc.services;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.Status;
+import io.grpc.channelz.v1.ChannelzGrpc;
+import io.grpc.channelz.v1.GetChannelRequest;
+import io.grpc.channelz.v1.GetChannelResponse;
+import io.grpc.channelz.v1.GetServerSocketsRequest;
+import io.grpc.channelz.v1.GetServerSocketsResponse;
+import io.grpc.channelz.v1.GetServersRequest;
+import io.grpc.channelz.v1.GetServersResponse;
+import io.grpc.channelz.v1.GetSocketRequest;
+import io.grpc.channelz.v1.GetSocketResponse;
+import io.grpc.channelz.v1.GetSubchannelRequest;
+import io.grpc.channelz.v1.GetSubchannelResponse;
+import io.grpc.channelz.v1.GetTopChannelsRequest;
+import io.grpc.channelz.v1.GetTopChannelsResponse;
+import io.grpc.internal.Channelz;
+import io.grpc.internal.Channelz.ChannelStats;
+import io.grpc.internal.Channelz.ServerList;
+import io.grpc.internal.Channelz.ServerStats;
+import io.grpc.internal.Channelz.SocketStats;
+import io.grpc.internal.Instrumented;
+import io.grpc.stub.StreamObserver;
+
+public class ChannelzService extends ChannelzGrpc.ChannelzImplBase {
+  private static final int DEFAULT_MAX_PAGE_SIZE = 25;
+  private final Channelz channelz;
+  private final int maxPageSize;
+
+  public ChannelzService() {
+    this(Channelz.instance(), DEFAULT_MAX_PAGE_SIZE);
+  }
+
+  @VisibleForTesting
+  ChannelzService(Channelz channelz, int maxPageSize) {
+    this.channelz = channelz;
+    this.maxPageSize = maxPageSize;
+  }
+
+  /** Returns top level channel aka {@link io.grpc.internal.ManagedChannelImpl}. */
+  @Override
+  public void getTopChannels(
+      GetTopChannelsRequest request, StreamObserver<GetTopChannelsResponse> responseObserver) {
+    try {
+      Channelz.RootChannelList rootChannels
+          = channelz.getRootChannels(request.getStartChannelId(), maxPageSize);
+      GetTopChannelsResponse.Builder responseBuilder = GetTopChannelsResponse
+          .newBuilder()
+          .setEnd(rootChannels.end);
+      for (Instrumented<ChannelStats> c : rootChannels.channels) {
+        responseBuilder.addChannel(ChannelzProtoUtil.toChannel(c));
+      }
+      responseObserver.onNext(responseBuilder.build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
+      responseObserver.onError(t);
+    }
+  }
+
+  /** Returns a top level channel aka {@link io.grpc.internal.ManagedChannelImpl}. */
+  @Override
+  public void getChannel(
+      GetChannelRequest request, StreamObserver<GetChannelResponse> responseObserver) {
+    try {
+      Instrumented<ChannelStats> s = channelz.getRootChannel(request.getChannelId());
+      if (s == null) {
+        responseObserver.onError(Status.NOT_FOUND.asRuntimeException());
+        return;
+      }
+
+      responseObserver.onNext(
+          GetChannelResponse
+              .newBuilder()
+              .setChannel(ChannelzProtoUtil.toChannel(s))
+              .build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
```

Comment:
same here.  It would be better to fail hard, rather than silently.

Reply:
done